### PR TITLE
Pytorch support for Join and Careduce Ops

### PR DESCRIPTION
**Pull request by HarshvirSandhu**
## Description
Add pytorch support for common CARdeuce Ops (sum, prod, all, any, ...)

## Related Issue

- [ ] Closes #
- [x] Related to #821 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/link/pytorch/dispatch/basic.py [hunk 1]
```diff
@@ -89,3 +89,20 @@ def arange(start, stop, step):
...
 
     return arange
+
+
+@pytorch_funcify.register(Join)
+def jax_funcify_Join(op, **kwargs):
+    def join(axis, *tensors):
+        # tensors could also be tuples, and in this case they don't have a ndim
+        tensors = [torch.tensor(tensor) for tensor in tensors]
+        view = op.view
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:99]: We wast to remove this kwarg, so you can just raise NotImplementedErrror if the Op has it set in the outer dispatch function.

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 2]
```diff
@@ -34,3 +35,11 @@ def dimshuffle(x):
         return res
 
     return dimshuffle
+
+
+@pytorch_funcify.register(Sum)
```
[the remark this answers is not in the document]
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/elemwise.py:40]: Yes, just added `Sum` for the initial commits.
Will also be adding `Prod`, `Any`, `All`

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 3]
```diff
@@ -37,6 +38,64 @@ def dimshuffle(x):
...
+
+    def torch_all(x):
+        return torch.all(x, dim=dim)
+
+    return torch_all
+
+
+@pytorch_funcify.register(Prod)
+def pytorch_funcify_prod(op, **kwargs):
+    dim = op.axis[0]
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:61]: Why axis[0]?

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 4]
```diff
@@ -37,6 +38,64 @@ def dimshuffle(x):
     return dimshuffle
 
 
+@pytorch_funcify.register(Sum)
+def pytorch_funcify_sum(op, **kwargs):
+    def torch_sum(x):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:43]: Nitpick, any reason why here you do op.axis, and in the other implementations you explicitly extract the axis? If this works fine, we can save lines in the other implementations?

File: tests/link/pytorch/test_elemwise.py [hunk 5]
```diff
@@ -57,6 +58,72 @@ def test_pytorch_elemwise():
     compare_pytorch_and_py(fg, [[0.9, 0.9]])
 
 
+@pytest.mark.parametrize("axis", [None, 0, 1])
```
Comment (ricardoV94) [inline tests/link/pytorch/test_elemwise.py:61]: Can we parametrize these tests with the reduce function? Since they all look the same, we can reduce a bunch of lines. Or at least separate only those that need numerical inputs from those that need boolean (all and any).

Comment (ricardoV94) [reply tests/link/pytorch/test_elemwise.py:61]: Also I would like to test axis = (1, 2), and have `a_pt` be a tensor3, so that we cover the case with more than 1 axis, but not all of them.

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pytensor/link/pytorch/dispatch/elemwise.py (hunk 3 above)
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/elemwise.py:61]: `op.axis` is a tuple, pytorch expects integers for `dim`

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py:61]: Then we need to change the logic, because it is possible for them to be tuples with more than one entry

File: pytensor/link/pytorch/dispatch/elemwise.py (hunk 4 above)
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/elemwise.py:43]: Sounds good. Also I noticed the Ops also have a `keepdims` argument.
But `op.keepdims` does not extract this

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py:43]: keepdims is actually not a property of the Op but a kwarg to the helper function. It just introduces an expand_dims (dimshuffle) on the output of CAReduce, so we don't have to do anything

File: pytensor/link/pytorch/dispatch/elemwise.py (hunk 3 above)
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/elemwise.py:61]: We could do something like this:
```
for d in op.axis:
     x=torch.prod(x, dim=d, keepdim=True) # To make sure constant shape, we can reshape in the end
                
for _ in op.axis:
     x.squeeze_(-1)
```

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py]: If you reduce in reversed order you don't have to worry about the keepdims. Sounds good, a bit surprising that they don't support multiple axes

File: tests/link/pytorch/test_elemwise.py [hunk 6]
```diff
@@ -79,113 +80,19 @@ def test_pytorch_sum(axis):
...
+    x = fn(a_pt, axis=axis)
     x_fg = FunctionGraph([a_pt], [x])
 
     compare_pytorch_and_py(x_fg, [test_value])
 
 
+@pytest.mark.parametrize("fn", [ptm.any, ptm.all])
 @pytest.mark.parametrize("axis", [None, 0, 1, (0, 1)])
-def test_pytorch_any(axis):
+def test_pytorch_any_all(fn, axis):
```
Comment (ricardoV94) [inline tests/link/pytorch/test_elemwise.py:91]:
```suggestion
def test_pytorch_careduce_bool(fn, axis):
```

File: tests/link/pytorch/test_elemwise.py [hunk 7]
```diff
@@ -58,8 +58,9 @@ def test_pytorch_elemwise():
     compare_pytorch_and_py(fg, [[0.9, 0.9]])
 
 
+@pytest.mark.parametrize("fn", [ptm.sum, ptm.prod, ptm.max, ptm.min])
 @pytest.mark.parametrize("axis", [0, 1, (0, 1), (1, 2), (1, -1)])
```
Comment (ricardoV94) [inline tests/link/pytorch/test_elemwise.py:62]: I think this is sufficient
```suggestion
@pytest.mark.parametrize("axis", [None, 0, 1, (0, -1)])
```

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 8]
```diff
@@ -58,9 +58,9 @@ def torch_all(x):
 def pytorch_funcify_prod(op, **kwargs):
     def torch_prod(x):
         if isinstance(op.axis, tuple):
-            for d in op.axis:
-                x = torch.prod(x, dim=d, keepdim=True)
-            return x.squeeze()
+            for d in op.axis[::-1]:
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:61]: More readable?
```suggestion
            for d in sorted(op.axis, reverse=True):
```

**Review by ricardoV94 — APPROVED** [review]
